### ISSUE TITLE
Publish consolidated item context

### ISSUE THREAD

**Issue by defagos** [SRGSSR/pillarbox-apple#559]
**As a Pillarbox developer** I want to build complex behaviors on consistent state. Currently essential item-related information (status, seekable ranges, media selection, etc.) stems from various asynchronous sources which might not deliver the same information at the same time, which can lead to subtle synchronization issues.

# Context

When implementing track analytics (#508) we discovered that the initial `play` event has undefined audio and subtitle selection information. While this is not an issue according to our specifications this is still puzzling.

The issue is related to the fact that our tracker subscribes to various sources of change, several of which are separately dependent on the item `status`. In many cases relevant information like the current media selection, for example, is only delivered once the item is ready to play. The item state publisher and these other publishers, all dependent on the item status, deliver their changes at different times (almost at the same time), leading to an inconsistent picture.

Thus we should attempt to:

- Group item-related properties that depend on the status in a common context struct (or similar).
- Deliver all these changes via the same pipeline, ensuring they are all updated in sync, especially when playback starts.
- Subscribe to relevant publishers so that the context is later updated, e.g. when the seekable range or the media selection change.

This should have no negative impact on our asynchronous / reactive approach but would increase state coherence. It would likely be a bit more challenging to test context updates globally but we can likely still test individual sources of change.

If we can rewrite our implementation with this strategy initial `play` events should readily have correct media selection information.

# Acceptance criteria

- The above strategy has been investigated (and implemented if possible).
- The API remains simple to use.
- No performance issue is added because of this change.

# Tasks

- [x] Identify related item state information which depends on the status being ready to play.
- [x] Implement single item context delivery pipeline.
- [x] Hook to a player current item context published property and write existing published properties, not connectable anymore, as computed properties.

**Comment by defagos** [comment]
Here are the item properties that currently require the item to be ready to play before being meaningful:

- time range (loaded + seekable)
- duration
- media selection
- presentation size

In fact these are all properties of interest for an item, it is therefore likely we should consolidate everything item-related.

**Comment by defagos** [comment]
I implemented a PoC with hierarchical state management on `559-publish-consolidated-item-context-poc`. This PoC works really well and fixes any kind of possible mismatches between pipelines, as all the state is consolidated into one. Some remarks:

- All the state is consolidated into a `context` (we can likely find a better name). We can publish this context, letting any kind of implementation subscribe to internal player updates with a single point of contact. This should make implementing custom trackers a breeze, or make it possible to easily implement a reactive UI without SwiftUI.
- The loaded and seekable time ranges should likely be loaded separately to keep updates minimal in the main state pipeline (~10 updates).
- This approach does not fix issues with media selection, though. We need to wait until the item is ready to play before we retrieve the media selection, which means that the initial playing state is notified before the selection has a chance to be loaded. But this approach should make it easy to fix this issue (if this is an issue) in the tracker directly.

Properly implementing this architecture is quite some work, though. We should:

- Avoid touching the existing code and introduce all hierarchical structures for storing state, migrating tests which make sense one by one.
- Once all tools are ready and tested we can hook the context delivery to a player published property. We can then replace prior published properties one by one with computed properties.
-

**Comment by defagos** [comment]
We have removed all the code we had to manage errors in playlists. Now back navigation will get stuck on failed items but this should be a corner case. In most cases the user will probably have another way of getting back to items prior to the failed one anyway (e.g. a playlist item selection interface).

For the moment we decided to keep things simple. We will improve the situation later if the need arises.

**Comment by defagos** [comment]
Regarding import optimization, some imports are superfluous when they involve a category (in this case any file in the same module can perform the import so that the code compiles). This is a [compiler bug](https://discord.com/channels/@me/1041985801443086368/1156530654482280489) but imports can still be optimized in the following way:

- Check minimal imports required.
- Look for public category method use and add non-required import (syntax coloring will then be fixed).

In some cases, e.g. where we have to import `Core` and `SRGDataProviderCombine` in the demo, we could use this fact to our advantage when only category methods were accessed. This allowed us to fix ambiguities without module prefixing but is fragile at best.

**Comment by defagos** [comment]
We have entirely revisited our reactive player implementation:

- All playback-related state is now consolidated using a single publisher into a `PlayerProperties`.
- Only some properties are seen as essential state (core properties) and lead to the `Player` publishing changes automatically. These properties can be directly read from a `Player` instance.
- Other properties can be read from a publisher which uses multicast so that each player instance only has one state publisher shared among all subscribers. This ensures consistent state throughout subscribers while avoiding costly publisher multiplication with each subscription.
- Local state extraction in SwiftUI views can be made with an `onReceive(player:at:to:)` modifier. This ensures that view hierarchies can be optimized so that frequent updates are only performed locally.

Some work remains to possibly consolidate more state (items and playback speed) but this has been deferred to other issues.
